Milestones: Add simulation based tests

manav2401 opened the pull request:
# Description

This PR adds automated simulation based tests for milestones. More info available [here](https://www.notion.so/polygontechnology/Testing-strategy-for-milestones-1b389b1249ae42a0a04a2e45e1f2f080). There are 2 primary tests. 
1. The base case which tries to perform a network partition of type 3:1 (as per stake) and rely on proposed milestones to follow the correct fork. 
2. Another case which tries to perform an equal network partition of type 1:1 (as per stake) and waits for milestones to get proposed on rejoin. 

Note that to reproduce the tests, you'd need a custom branch of milestone on bor (one which fixes the primary validator). 

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [x] New test case for remote devnet

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

#### It should never be the case...

- [x] This PR requires changes to bor
  - PR [here](https://github.com/maticnetwork/bor/pull/654)
- [x] This PR requires changes to heimdall
  - PR [here](https://github.com/maticnetwork/heimdall/pull/939):

## Testing

- [ ] I have tested this code manually on local environment
- [x] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai

temaniarpit27 commented on src/express/commands/milestone.js:
```diff
@@ -0,0 +1,400 @@
+  runSshCommandWithReturn
+} = require('../common/remote-worker')
+
+const milestoneLength = 64
```
check for this value

manav2401 replied:
For devnets in matic-cli it's still 64. I remember Vaibhav mentioning that it's reduced to 12 now, but the milestone which are proposed are still using 64.

temaniarpit27 replied:
@VAIBHAVJINDAL3012 please check this

VAIBHAVJINDAL3012 replied:
Actually Manav forgot to take the pull, We discussed over it.

temaniarpit27 commented on src/express/commands/milestone.js:
```diff
@@ -0,0 +1,400 @@
+async function removePeers(ip, peers) {
+  let tasks = []
+  for (let i = 0; i < peers.length; i++) {
+    let command = `~/go/bin/bor attach ~/.bor/data/bor.ipc --exec "admin.removePeer('${peers[i]}')"`
```
shouldn't the paths be according to v0.3?
right now this may work but it should be made to new paths so that in future we don't need to revisit this

manav2401 replied:
Matic cli still uses the old paths and I saw other commands using the same so didn't change it. Maybe once all things are working we can create a global path variable for bor path and use it everywhere instead of hardcoding it.

temaniarpit27 commented on src/express/commands/milestone.js:
```diff
@@ -0,0 +1,400 @@
+async function addPeers(ip, peers) {
+  let tasks = []
+  for (let i = 0; i < peers.length; i++) {
+    let command = `~/go/bin/bor attach ~/.bor/data/bor.ipc --exec "admin.addPeer('${peers[i]}')"`
```
same here

temaniarpit27 commented on src/express/commands/milestone.js:
```diff
@@ -0,0 +1,400 @@
+}
+
+async function getPeerLength(ip) {
+  const command = `/home/ubuntu/go/bin/bor attach ~/.bor/data/bor.ipc --exec "admin.peers.length"`
```
same here

temaniarpit27 commented on src/express/commands/milestone.js:
```diff
@@ -0,0 +1,400 @@
+
+  // Check for number of validators
+  if (doc['numOfValidators'].length < 4) {
+    console.log('📍Cannot run milestone tests on less than 4 validator nodes')
```
can you explain why we need >=4 vals?

manav2401 replied:
Just for simplicity of the tests. If the number of validators are 3 (or less), we won't get group with majority votes while creating clusters. As it's going to be used only internally, I added this check.

marcello33 commented on src/express/commands/monitor.js:
```diff
@@ -58,6 +58,12 @@ async function checkCheckpoint(ip) {
+export async function checkLatestMilestone(ip) {
+  const url = `http://${ip}:1317/milestone/latest`
+  const response = await fetch(url)
+  return await response.json()
```
Not sure here tbh, but are all the `await` necessary?
I see we are awaiting on `fetch(url)`, `response.json()` and when the function itself is called `milestone = await checkLatestMilestone(borHosts[0])`

manav2401 replied:
I think they're required. For fetch, it's a network call so we'll have to wait. `.json()` function returns a promise and does not return the required result immediately hence we'd need an await there. Moreover, due to this the function `checkLatestMilestone` returns a promise so we'll have to add an await before calling it. Wdyt?

marcello33 replied:
Yeah, right, makes sense, thanks

marcello33 commented on src/express/common/milestone-utils.js:
```diff
@@ -0,0 +1,396 @@
+}
+
+export async function createClusters(ips, enodes, split = 1) {
+  // `split` defines how clusters are created and which index to use to seperate nodes.
```
typo in `seperate `

marcello33 commented on src/express/common/milestone-utils.js:
```diff
@@ -0,0 +1,396 @@
+  return response
+}
+
+export async function rejoinClusters(ips, enodes, split = 1) {
```
`rejoinClusters` and `validateNumberOfPeers` are not used, right?
In case, shall we remove them?

marcello33 commented on src/express/common/milestone-utils.js:
```diff
@@ -0,0 +1,396 @@
+    finalizedBlocks = values
+  })
+
+  // if (finalizedBlocks.length != hosts) {
```
shall we remove this comment?

marcello33 commented on src/express/common/milestone-utils.js:
```diff
@@ -0,0 +1,396 @@
+  return true
+}
+
+export async function createClusters(ips, enodes, split = 1) {
```
Is this function the only one needed to create a deterministic fork on the devnet?
I know it's out of scope for your PR, but was wondering if I can creare a command for `express-cli` out of it, something like `--reorg [split]`

manav2401 replied:
Oh interesting. Yeah we could actually do so.

marcello33 replied:
Nice, thanks!
If you want, once other comments are addressed, I can work on it and then submit to you for review

marcello33 commented on src/express/common/remote-worker.js:
```diff
@@ -150,3 +150,22 @@ export async function runScpCommand(src, dest, retries) {
+    return response
+  } catch (error) {
+    if (retries - 1 > 0) {
+      console.log('--- running command')
```
Shall we remove this console log, just to be consistent with other similar functions?

manav2401 replied:
Sorry missed this. Will remove.

marcello33 commented on src/express/common/milestone-utils.js:
```diff
@@ -0,0 +1,396 @@
+
+  const responseJson = await response.json()
+  if (responseJson.result) {
+    // console.log(`📍Request. number: ${number}, opts: ${JSON.stringify(opts)}`)
```
Shall we remove this comment?

manav2401 commented on src/express/commands/milestone-base.js:
```diff
@@ -0,0 +1,341 @@
+const queryTimer = (milestoneLength / 4) * 1000
+
+export async function milestoneBase() {
+  // NOTE: Make sure bor branch has logic for hardcoded primary validator
```
So, in order to run deterministic tests for milestones, we need to make sure that in the devnet setup when we induce a reorg, there are 2 partitions where one of them has higher difficulty but less voting power and the other one has lower difficulty but more voting power. 

If we run this with the default setup of devnet, this situation may or may not arise. Hence we have to hardcode the primary validator in all the nodes so that we know which fork will have the highest difficulty. This speeds up the process of testing as we're bound to get results on every run.

manav2401 commented on src/express/commands/milestone-base.js:
```diff
@@ -0,0 +1,341 @@
+    enodes = values
+  })
+
+  if (enodes.includes('')) {
```
The function `getEnode` returns an empty string if there's an error in making the `admin.nodeInfo.enode` call. Hence, before proceeding with the tests, we're making sure that we have received enode of all the nodes.

manav2401 commented on src/express/commands/milestone-base.js:
```diff
@@ -0,0 +1,341 @@
+  }
+
+  console.log('📍Waiting 10s to fetch finalized blocks...')
+  await timer(10000)
```
The timer to query milestone is quite low for bor and hence if a milestone just got proposed, we're sure that we'll get a finalised block. This can be made configurable via some global params tho.

manav2401 replied:
If it doesn't have the milestone, the function below should give an error. We shouldn't retry IMO as something went wrong with saving milestones.

manav2401 commented on src/express/commands/milestone-base.js:
```diff
@@ -0,0 +1,341 @@
+const milestoneLength = 64
+const queryTimer = (milestoneLength / 4) * 1000
+
+export async function milestoneBase() {
```
Yeah we could. I wasn't sure how much times we'll be conducting these tests so didn't put much efforts on that. I will try to abstract some logic into functions in form of steps.

manav2401 commented on src/express/commands/milestone-helper.js:
```diff
@@ -0,0 +1,221 @@
+/* eslint-disable dot-notation */
```
I'd keep the `getMiner` one as it might be helpful in some future tests (not necessarily for milestones). Will remove the rest.

marcello33 commented:
Also, @manav2401 can you include the instructions on how to run such tests in the `README`? I'll make sure to have them also in the epic description on notion. Thanks.

manav2401 commented:
Have addressed all the comments, refactored the changes by a great extent and updated all the steps in a readme. Looking for review @marcello33. Thanks!

marcello33 commented:
Thanks @manav2401, LGTM!
I just fixed the linter and pushed the implementation of the `--reorg [index]` command.
Can you please double check it does what it is supposed to do?
I have some doubts on the `validateClusters` part, around the log
`console.log(`Peer length mismatch, got: ${peers}, expected: ${expectedPeers}`)`
Will this work with any custom `index` the way I implemented it?
Thanks

manav2401 commented:
Thanks for the feature changes @marcello33. Yes given the implementation, it should work for any `index` as the expected peers is calculated based on that only. 

Also, few more things. 
1. Do you think we should quickly add a `joinPeers` command so that once the reorg is done, one can directly re-connect all the nodes?
2. I am not fully sure of the wordings used for this command in README. The parameter `index` does not technically represent the index of the node in list of nodes. It basically represents the number of nodes that one of the clusters will have (with other being total number of nodes - index). Do you think we should change it?

manav2401 commented:
Thanks @marcello33 
LGTM!
